# Aim : To Study the impact of different `Callbacks`.

**Data Preprocessing**

In [1]:
# TensorFlow and tf.keras
import tensorflow as tf

# Helper libraries
import numpy as np
from tensorflow.keras import initializers
from tensorflow.python.keras import activations

print(tf.__version__)

# downloading fashion_mnist data
fashion_mnist = tf.keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

train_images = train_images / 255.0

test_images = test_images / 255.0       


c:\users\sonu.ramkumar.jha\desktop\experiments\env\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
c:\users\sonu.ramkumar.jha\desktop\experiments\env\lib\site-packages\numpy\.libs\libopenblas.GK7GX5KEQ4F6UYO3P26ULGBQYHGQO7J4.gfortran-win_amd64.dll
c:\users\sonu.ramkumar.jha\desktop\experiments\env\lib\site-packages\numpy\.libs\libopenblas.WCDJNK7YVMPZQ2ME2ZZHJJRJ3JIKNDB7.gfortran-win_amd64.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


2.5.0


**Preparing Model**

In [2]:
activation = tf.keras.activations.tanh

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])


model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


**EarlyStopping Callbacks**

In [3]:
# callbacks
es = tf.keras.callbacks.EarlyStopping(patience=1, monitor='accuracy')

model.fit(train_images, train_labels, epochs=100, callbacks=[es])

test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('test_loss', test_loss)
print('test_accuracy', test_acc)

Epoch 1/100
1875/1875 [==============================] - 3s 1ms/step - loss: 0.4759 - accuracy: 0.8296
Epoch 2/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3693 - accuracy: 0.8642
Epoch 3/100
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3350 - accuracy: 0.8765
Epoch 4/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3136 - accuracy: 0.8846
Epoch 5/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2956 - accuracy: 0.8901
Epoch 6/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2841 - accuracy: 0.8944
Epoch 7/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2704 - accuracy: 0.8992
Epoch 8/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2595 - accuracy: 0.9039
Epoch 9/100
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2497 - accuracy: 0.9067
Epoch 10/100
1875/1875 [==============================] - 3s 1ms/step - l

**Observations**
- As you can see the model stopped training after 22 epochs since the last two accuracies 0.933 and 0.933 as same.
- In the absence of EarlyStopping it would have run till 100 epochs.
- Advantage of `EarlyStopping` is you can set the stopper based on some critrian you choose.

**ModelCheckpoint Callback**

In [8]:
# deleting previous model
del model

# new model
activation = tf.keras.activations.tanh

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])


model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_2 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_5 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


In [9]:
# save the best model based on train accuracy to the path model/checkpoint 
mc = tf.keras.callbacks.ModelCheckpoint(filepath='model/checkpoint', save_best_only=True, monitor='accuracy')

In [10]:
model.fit(train_images, train_labels, epochs=30)
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2, callbacks=[mc])

print('test_loss', test_loss)
print('test_accuracy', test_acc)

Epoch 1/30
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4765 - accuracy: 0.8291
Epoch 2/30
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3671 - accuracy: 0.8656
Epoch 3/30
1710/1875 [==========================>...] - ETA: 0s - loss: 0.3311 - accuracy: 0.8795

KeyboardInterrupt: 

`Here after running the three epochs, I stopped the training. Now, because of the model checkpoint it should automatially load the previous best model and train that model instead of training from scratch.`

In [12]:
model.fit(train_images, train_labels, epochs=10)
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2, callbacks=[mc])

print('test_loss', test_loss)
print('test_accuracy', test_acc)

Epoch 1/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3153 - accuracy: 0.8841
Epoch 2/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2966 - accuracy: 0.8897
Epoch 3/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2801 - accuracy: 0.8968
Epoch 4/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2695 - accuracy: 0.9008
Epoch 5/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2586 - accuracy: 0.9053
Epoch 6/10
1875/1875 [==============================] - ETA: 0s - loss: 0.2495 - accuracy: 0.90 - 3s 1ms/step - loss: 0.2494 - accuracy: 0.9069
Epoch 7/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2422 - accuracy: 0.9105
Epoch 8/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2324 - accuracy: 0.9143
Epoch 9/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2285 - accuracy: 0.9144
Epoch 10/10
1875/1875 [=================

`As you can see that model has started training at loss:0.3153  and accuracy : 0.8841, which is just after in intrupted epoch where loss :0.3311 and accuracy is 0.8795. `

**Observations**

- It is used to save the model based on critrian you choose. 
- In case the training intrupts it will save the model automatically and for the next time we don't have to train from scratch.